src/token/cose/recipient/mod.rs

```diff
@@ -0,0 +1,832 @@
+    /// or otherwise unsuitable for decryption, this function will return the most fitting
+    /// [`CoseCipherError`] for the specific type of error.
+    ///
+    /// If Additional Authenticated Data is provided even though the chosen algorithm is not an AEAD
+    /// algorithm, a [`CoseCipherError::AadUnsupported`] will be returned.
+    ///
+    /// If the COSE object is not malformed, but an error in the cryptographic backend occurs, a
+    /// [`CoseCipherError::Other`] containing the backend error will be returned.
+    /// Refer to the backend module's documentation for information on the possible errors that may
+    /// occur.
+    ///
+    /// If the COSE object is not malformed, but signature verification fails for all key candidates
+    /// provided by the key provider a [`CoseCipherError::NoMatchingKeyFound`] error will be returned.
+    ///
+    /// The error will then contain a list of attempted keys and the corresponding error that led to
+    /// the verification error for that key.
+    /// For an invalid MAC for an otherwise valid and suitable object+key pairing, this would
+    /// usually be a [`CoseCipherError::VerificationFailure`].
+    ///
+    /// # Examples
+    ///
+    /// Refer to the trait-level documentation for an example.
+    fn try_decrypt<B: KeyDistributionCryptoBackend, CKP: KeyProvider, CAP: AadProvider + ?Sized>(
+        &self,
+        backend: &mut B,
+        key_provider: &CKP,
+
+        context: EncryptionContext,
+        external_aad: CAP,
+    ) -> Result<Vec<CoseKey>, CoseCipherError<B::Error>>;
+}
+
+impl CoseRecipientExt for CoseRecipient {
+    fn try_decrypt<B: KeyDistributionCryptoBackend, CKP: KeyProvider, CAP: AadProvider + ?Sized>(
+        &self,
+        backend: &mut B,
+        key_provider: &CKP,
+        context: EncryptionContext,
+        external_aad: CAP,
+    ) -> Result<Vec<CoseKey>, CoseCipherError<B::Error>> {
+        let alg = match determine_algorithm::<B::Error>(
+            None,
+            Some(&self.protected.header),
+            Some(&self.unprotected),
+        ) {
+            Ok(v) => v,
+            Err(e) => {
+                // A CoseRecipient MUST always have an algorithm set (see RFC 9052,
+                // Section 8), which means that this COSE object is malformed.
+                // TODO in theory, we could proceed here by skipping over this recipient,
+                //      but I'm not sure if we should continue parsing a
+                //      non-standard-compliant object...
+                return Err(e);
+            }
+        };
+
+        // Determine key operations that fulfill the requirements of the algorithm.
+        let operation = determine_decrypt_key_ops_for_alg(alg)?;
+
+        // Direct => Key of key provider will be used for lower layer directly.
+        // TODO ensure that Direct is the only method used on the message (RFC 9052, Section 8.5.1)
```

Comment:
The issue here is that in order to perform this check reliably, we would have to iterate over the entire tree once before even starting the search for recipient candidates, which might have performance implications (it would basically require an additional traversal of the entire Tree).
I'll create a new issue for this and reference it here.

Reply:
Yes, it's also fine for us not to support this, though in that case we should make it clear in the documentation.

Reply:
Maybe could be something worth considering for a strict mode.

For now, I will amend the documentation for the nested operations to indicate that this check is not done.